README
Net-Dynect-REST version 0.01
============================

Dynect.net is a provider of hosted DNS services. They offer a /two/ 
APIs for interacting with their service at this stage - REST and SOAP. 
This set of modules attempts to implement (some of) the REST interface.

BASE MODULES
------------
There are three base modules to use:
  Net::Dynect::REST - will hold your session information (Auth Token)
  Net::Dynect::REST::Request - will format a request and parameters
  Net::Dynect::REST::Response - will decode the response

To use these, instantiate the Net::Dynect::REST with your security 
credentials. Then create a request object, and pass  it to the 
Net::Dynect::REST->execute() method, and you should get a 
return value of Net::Dynect::REST::Response.

  my $dynect = Net::Dynect::REST->new(user_name => 'me', 
				      customer_name => 'my_company', 
				      password => 'my_secret');
  my $request = Net::Dynect::REST::Request->new(..your params..);
  my $response = $dynect->execute($request);
  .. time passes... you do work...
  $dynect->logout;



ASBTRCTION
----------
Then there are a set of higher level abstractions modules:
  Net::Dynect::REST::Zone - a DNS Zone
  Net::Dynect::REST::ZoneChanges - the pending changes
  Net::Dynect::REST::ARecord - an (IPv4) Address resource record
  ....

To use these, you need to:
  1. Create your Net::Dynect::REST session as above.
  2. Create your asbtration object, such as "Zone", and pass it your 
     Net::Dynect::REST.
  3. Use the methods on the abstraction object to do what you want 
     (add/remove/delete).
  4. If you're happy with all your changes, then you need to load your 
     Zone (if you hadn't already) and use the 'publish' method.

For example:
  my $dynect = Net::Dynect::REST->new(user_name => 'me', 
				      customer_name => 'my_company', 
				      password => 'my_secret');
  my $zone = Net::Dynect::REST::Zone->new(connection => $dynect,
					  zone => 'example.com');
  my @addresses = Net::Dynect::REST::ARecord->find(connection => $dynect,
					zone => 'example.com',
				        fqdn => 'foo.example.com');
  foreach (@addresses) {
    printf "foo.example.com = %s\n", $_->rdata->address;
  }

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  LWP - the Perl web library
  Carp - for various wanrings, normally already in your base Perl
  JSON - for encoding and decoding data

THANKS

This is only possible thanks to LWP, JSON, some inspiraction from DBI 
(Tim Bunce), general inspiration over the years from Acme (Léon 
Brocard), a random post from Mark Stosberg on the CGIAPP mailing list.
Also, thanks to NetShelter and Dynect.

COPYRIGHT AND LICENCE

Copyright (C) 2010 by James Bromberger

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Net::Dynect::REST.

0.07  Mon Sep 27 15:30:00 2010
	- Start to handle any deferred responses that throw a Job reference. You must check the ref() of the response to ensure you have a response object (Net::Dynect::REST::Response), and not a Job object (Net::Dynect::REST::Job). If you get a job object, then you need to check back on $job->find() to get the eventual Response object, when it is ready.

0.06  Mon Sep 27 13:45:00 2010
	- Fix another few bugs in Net::Dynect::REST::ResourceRedord->delete().
	- Add explicit dependency on HTTP::Request::Common version 5.814 or newer to support DELETE.
	- Fix typos in documentation in MXRecord.pm and RData.pm.

0.05  Mon Sep 27 13:30:00 2010
	- Fix bug in Net::Dynect::REST::ResourceRedord->delete().

0.04  Mon Sep 27 12:25:00 2010
	- Drop required Perl from 5.10.1 to 5.8.0 to support some more older systems.

0.03  Mon Sep 27 10:00:00 2010
	- add in dependencies so module will pass tests! *sigh*

0.02  Sat Sep 25 09:35:00 2010
	- update MANIFEST to include the rest of the code.
	- update documentation references in many modules.

0.01  Fri Sep 24 09:14:26 2010
	- original version; created by h2xs 1.23 with options
		-AX -n Net::Dynect::REST
	- implements Net::Dynect::REST::Request and Net::Dynect::REST::Response for use with Net::Dynect::REST, and a select few objects to use this API: ARecord, Zone, and ZoneChanges. This gives the ability to add/delete/update A Records ina  zone and publish it.


lib/Net/Dynect/REST.pm
package Net::Dynect::REST;
# $Id: REST.pm 149 2010-09-26 01:33:15Z james $
use strict;
use warnings;
use Net::Dynect::REST::Request;
use Net::Dynect::REST::Response;
use Net::Dynect::REST::Session;
use LWP::UserAgent;
use HTTP::Request::Common;
use Time::HiRes qw(gettimeofday tv_interval);
use Carp qw(carp cluck);
our $VERSION = do { my @r = (q$Revision: 149 $ =~ /\d+/g); sprintf "%d."."%03d" x $#r, @r };

=head1 NAME

Net::Dynect::REST - A REST implementation to communicate with Dynect

=head1 SYNOPSIS

 use Net::Dynect::REST
 my $dynect = Net::Dynect::REST->new();
 $dynect->login(user_name => $user, customer_name => $customer, password => $password;

=head1 METHODS

=head2 Creating

=over 4

=item Net::Dynect::REST->new()

This constructor will return an object, and can optionally attempt to establish a session if sufficient authentication details are passed as parameters. It takes the optional arguments of:

=over 4

=item * debug

A numeric debug level, where 0 is silent, 1 is standard output, and higher gives more details.

=item * server

=item * protocol

=item * base_path

=item * port

=item * user_name

=item * customer_name

=item * password

=over 4

=back

=back

=cut

sub new {
    my $proto = shift;
    my $self  = bless {}, ref($proto) || $proto;
    my %args  = @_;
    $self->_debug_level( $args{debug} )  if defined $args{debug};
    $self->server( $args{server} )       if defined $args{server};
    $self->protocol( $args{protocol} )   if defined $args{protocol};
    $self->base_path( $args{base_path} ) if defined $args{base_path};
    $self->port( $args{port} )           if defined $args{port};
    if (   defined( $args{user_name} )
        && defined( $args{password} )
        && defined( $args{customer_name} ) )
    {
        my $login = $self->login(
            user_name     => $args{user_name},
            customer_name => $args{customer_name},
            password      => $args{password}
        );
        if ( not $login ) {

            #carp "Tried to log in, but failed";
            return;
        }
    }
    return $self;
}

=back

=head2 Methods

=over 4

=item $dynect->login()

This will attempt to create a valid Session object by forming and sending a login request, and parsing the response. Parameters are:

=over 4

=item * user_name

=item * customer_name

=item * password

=back

=cut

sub login {
    my $self = shift;
    my %args = @_;
    if (
        not(   defined( $args{user_name} )
            && defined( $args{customer_name} )
            && defined( $args{password} ) )
      )
    {
        carp "Login method requires user_name, customer_name and password";
        return;
    }

    my $dynect_rest_request = Net::Dynect::REST::Request->new(
        operation => 'create',
        service   => 'Session',
        params    => {
            user_name     => $args{user_name},
            customer_name => $args{customer_name},
            password      => $args{password}
        }
    );

    if ( not $dynect_rest_request ) {
        carp "Invalid request object";
        return;
    }

    my $dynect_rest_response = $self->execute($dynect_rest_request);

    if ( not $dynect_rest_response ) {
        carp "Did not get a response object";
        return;
    }

    if ( $dynect_rest_response->status !~ /^success$/i ) {
        carp join( ', ', map { $_->info } @{ $dynect_rest_response->msgs } );
        return;
    }

    $self->session(
        Net::Dynect::REST::Session->new( response => $dynect_rest_response ) );
    return 1;
}

=item $dynect->logout()

If we have a valid session, then this will try and perform a logout against Dynect, and remove our sesssion object.

=cut

sub logout {
    my $self = shift;
    if ( not $self->session ) {
        carp "Cannot logout with out current session";
        return;
    }

    my $dynect_rest_request = Net::Dynect::REST::Request->new(
        operation => 'delete',
        service   => 'Session'
    );
    if ( not $dynect_rest_request ) {
        carp "Invalid request object";
        return;
    }

    my $dynect_rest_response = $self->execute($dynect_rest_request);

    if ( not $dynect_rest_response ) {
        carp "Did not get a response";
        return;
    }

    if ( $dynect_rest_response->status !~ /^success$/i ) {
        carp "Could not log out";
        return;
    }
    $self->session(undef);
    return 1;
}

=item $dynect->execute()

This is the main heavy lifting; where Net::Dynect::REST::Request objects get sent to the server, and a Net::Dynect::REST::Response is returned, if all is OK. It takes one argument - the Net::Dynect::REST::Request object.

=cut

sub execute {
    my $self                = shift;
    my $dynect_rest_request = shift;
    if ( ref($dynect_rest_request) ne "Net::Dynect::REST::Request" ) {
        carp "Need a request to execute";
        return;
    }
    elsif (
        not(   defined( $dynect_rest_request->service )
            && defined( $dynect_rest_request->operation ) )
      )
    {
        carp "Error with request - need to set operation and service: "
          . $dynect_rest_request;
        return;
    }

    my $uri = $self->base_uri . $dynect_rest_request->service . "/";

    my $http_request;
    if ( $dynect_rest_request->operation eq "create" ) {
        if ( $dynect_rest_request->params ) {
            $http_request = HTTP::Request::Common::POST(
                $uri,
                'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8',
                Content => $dynect_rest_request->params
            );
        }
        else {
            $http_request = HTTP::Request::Common::POST( $uri,
                    'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8' );
        }
    }
    elsif ( $dynect_rest_request->operation eq "read" ) {
        if ( $dynect_rest_request->params ) {
            $http_request = HTTP::Request::Common::GET(
                $uri,
                'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8',
                Content => $dynect_rest_request->params
            );
        }
        else {
            $http_request = HTTP::Request::Common::GET( $uri,
                    'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8' );
        }
    }
    elsif ( $dynect_rest_request->operation eq "update" ) {
        if ( $dynect_rest_request->params ) {
            $http_request = HTTP::Request::Common::PUT(
                $uri,
                'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8',
                Content => $dynect_rest_request->params
            );
        }
        else {
            $http_request = HTTP::Request::Common::PUT( $uri,
                    'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8' );
        }
    }
    elsif ( $dynect_rest_request->operation eq "delete" ) {
        if ( $dynect_rest_request->params ) {
            $http_request = HTTP::Request::Common::DELETE(
                $uri,
                'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8',
                Content => $dynect_rest_request->params
            );
        }
        else {
            $http_request = HTTP::Request::Common::DELETE( $uri,
                    'Content-Type' => $dynect_rest_request->mime_type
                  . '; charset=utf-8' );
        }
    }
    else {
        die "Unrecognised operation: " . $dynect_rest_request->operation;
    }

    $self->_debug( 4, "Making a request to $uri" );
    $self->_debug( 5,
        "Request will be:\n" . $dynect_rest_request . "\n" . ( '-' x 70 ) );

    my $time_start = [gettimeofday];

    my $http_response = $self->_webclient->request($http_request);

    my $time_elapsed         = tv_interval($time_start);
    my $dynect_rest_response = Net::Dynect::REST::Response->new(
        content          => $http_response->decoded_content,
        format           => $dynect_rest_request->format,
        request_duration => $time_elapsed,
        request_time     => $time_start->[0]
    );

    $self->_debug( 2,
        join( '; ', map { $_->info } @{ $dynect_rest_response->msgs } ) ) if defined $dynect_rest_response->msgs;
    $self->_debug( 5,
            "Response received in $time_elapsed:\n"
          . $dynect_rest_response . "\n"
          . ( '-' x 70 ) );

    return $dynect_rest_response;
}

=item $dynect->session()

This is a Net::Dynect::REST::Session object, which should eb the current valid session for this Net::Dynect::REST object to use. It updates the web client to include the B<Auth-Token> header for subsequent requests

=cut

sub session {
    my $self = shift;
    if (@_) {
        my $new = shift;
        if ( defined($new) && ref($new) ne "Net::Dynect::REST::Session" ) {
            carp "Invalid session: $new";
            return;
        }

        $self->{session} = $new;
        return
          unless
            defined $new;   # Allow the sesison to be undefined if its now dead!
        $self->_debug( 6, "Adding auth token to default headers" );
        $self->_webclient->default_headers(
            HTTP::Headers->new( ':Auth-Token' => $new->token ) );
    }
    return $self->{session};
}

sub _webclient {
    my $self = shift;
    if ( not defined $self->{_webclient} ) {
        $self->{_webclient} = LWP::UserAgent->new(
            agent     => ref($self) . "/" . $VERSION,
            env_proxy => 1
        );
    }
    return $self->{_webclient};
}

sub _debug_level {
    my $self = shift;
    if (@_) {
        my $new = shift;
        if ( $new !~ /^\d$/ ) {
            carp "Invalid debug level: " . $new;
            return;
        }
        $self->{_debug_level} = $new;
        $self->_debug( 0, "Debug set to " . $self->{_debug_level} );
    }
    return $self->{_debug_level} || 0;
}

sub _debug {
    my $self = shift;
    my ( $level, $message ) = @_;
    return unless $level <= $self->_debug_level;

    if ( $level > 8 ) {
        $Carp::CarpLevel = 1;
        cluck $message;
    }
    else {
        carp $message;
    }
}

=back

=head2 Attributes

=over 4

=item $dynect->server()

This is the server host name that we will send our requests to. Default is B<api2.dynect.net>.

=cut

sub server {
    my $self = shift;
    if (@_) {
        my $new = shift;
        $self->{server} = $new;
    }
    return $self->{server} || 'api2.dynect.net';
}

=item $dynect->protocol()

This is the protocol we will use, either B<http> or B<https>. Default is B<https>.

=cut

sub protocol {
    my $self = shift;
    if (@_) {
        my $new = shift;
        return unless $new =~ /^https?$/;
        $self->{protocol} = $new;
    }
    return $self->{protocol} || "https";
}

=item $dynect->base_path()

This is the path that is used to find the services we will be accessing. Default is B</REST/>.

=cut

sub base_path {
    my $self = shift;
    if (@_) {
        my $new = shift;
        return unless $new =~ m!^/[\w\d/-]*$!;
        $self->{base_path} = $new;
    }
    return $self->{base_path} || '/REST/';
}

=item $dynect->port()

The TCP port that we will use. The default is to use whatever is apropriate for the protocol.

=cut

sub port {
    my $self = shift;
    if (@_) {
        my $new = shift;
        return unless $new =~ /^\d+$/;
        $self->{port} = $new;
    }
    return $self->{port};
}

=item $dynect->base_uri()

A convenience method to put together the protocl, server, port and base_path attributes into a URI.

=cut

sub base_uri {
    my $self = shift;
    return unless ( $self->protocol && $self->server && $self->base_path );
    return
        $self->protocol . "://"
      . $self->server
      . ( defined( $self->port ) ? ":" . $self->port : "" )
      . $self->base_path;
}

=back

=head1 SEE ALSO

L<Net::Dynect::REST::Request>, L<Net::Dynect::REST::Response>, L<Net::Dynect::REST::info>.

=head1 AUTHOR

James bromberger, james@rcpt.to

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 by James Bromberger

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
